avro/src/error.rs

```diff
@@ -21,8 +21,1154 @@ use crate::{
 };
 use std::{error::Error as _, fmt};
 
+/// Errors encounterd by Avro.
+///
+/// To inspect the details of the error use [`details`](Self::details) or [`into_details`](Self::into_details)
+/// to get a [`Details`] which contains more precise error information.
+///
+/// See [`Details`] for all possible errors.
+#[derive(thiserror::Error, Debug)]
+#[repr(transparent)]
+#[error(transparent)]
+pub struct Error {
+    details: Box<Details>,
+}
+
+impl Error {
+    pub fn new(details: Details) -> Self {
+        Self {
+            details: Box::new(details),
+        }
+    }
+
+    pub fn details(&self) -> &Details {
+        &self.details
+    }
+
+    pub fn into_details(self) -> Details {
+        *self.details
+    }
+}
+
+/// Functions for constructing a specific error type.
+#[allow(non_snake_case, reason = "Want to mimic the `Details` variants")]
+impl Error {
+    /// Construct a new [`Error`] with a [`Details::SnappyCrc32`].
+    pub fn SnappyCrc32(expected: u32, actual: u32) -> Self {
```

Comment:
I wonder whether we really need to have these public factory methods.
Usually the library creates the "instance". The user applications check/match the error.
But even if the user applications need to create an instance they could replace `Error::Xyz{...}` with `Details::Xyz{...}.into()`, i.e. `s/Error/Details` and append `.into()`. We could use the same in the library too.

Reply:
I agree they don't need to be public.

We could remove them completely but then a lot more lines will have to be changed, if you want that let me know and I'll do it.

Reply:
I've marked the construction functions as `pub(crate)` for now, let me know if you want to completely remove them.